File: fixtures/lifecycle-shared-identity-with-no-accountable-principal/verify.py
#!/usr/bin/env python3
# Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
#
# Independent Python runner for the
# lifecycle-shared-identity-with-no-accountable-principal candidate family.
#
# Written from README.md and vectors.json rather than ported from harness.ts.
# The accountability rule here is a sequence of guard clauses over pre-filtered
# lists rather than the flag-parameterized function the TypeScript harness uses,
# so the two implementations agreeing is evidence about the rules.
#
# The only third-party import is the published PyPI agent-passport-system, for
# its RFC 8785 (JCS) canonicalization primitive.
#
# Run: /path/to/venv/bin/python \
#        fixtures/lifecycle-shared-identity-with-no-accountable-principal/verify.py
#
# Exit 0 on success, 1 on any failure.

from __future__ import annotations

import hashlib
import json
import sys
from datetime import datetime, timezone
from pathlib import Path

from agent_passport.canonical import canonicalize_jcs

HERE = Path(__file__).resolve().parent
VOCABULARY = {"valid", "invalid", "not_established", "not_yet_effective", "suspended", "restricted"}
ANOMALIES = {"not_applicable", "on_enumerated_list", "off_enumerated_list"}

problems: list[str] = []


def fail(message: str) -> None:
    problems.append(message)
    print(f"  FAIL {message}", file=sys.stderr)


def at(iso: str) -> datetime:
    return datetime.fromisoformat(iso.replace("Z", "+00:00")).astimezone(timezone.utc)


def anomaly_for(inp: dict) -> str:
    if inp["identity"]["kind"] != "privileged_unscoped":
        return "not_applicable"
    on_list = inp["action"]["task"] in inp["privileged_task_list"]
    return "on_enumerated_list" if on_list else "off_enumerated_list"


def accountability(inp: dict, *, check_interval: bool, check_multiplicity: bool) -> tuple[str, str]:
    if inp["identity"]["kind"] == "individual":
        return "valid", "individual_identity_bound"

    with_standing = [c for c in inp["checkouts"] if c["signer_has_standing"]]
    action_at = at(inp["action"]["at"])
    if check_interval:
        covering = [c for c in with_standing if at(c["from"]) <= action_at <= at(c["to"])]
    else:
        covering = list(with_standing)

    if len(covering) == 1:
        return "valid", "individual_bound_by_checkout"
    if len(covering) > 1:
        if check_multiplicity:
            return "not_established", "overlapping_checkouts"
        return "valid", "individual_bound_by_checkout"
    if with_standing:
        return "not_established", "no_checkout_covers_the_action"

    attributions = inp["attribution_records"]
    if not attributions:
        return "not_established", "shared_identity_attribution_pending_or_external"
    referencing = [a for a in attributions if a["references_record_sha256"] == inp["record_sha256"]]
    if not referencing:
        return "not_established", "attribution_does_not_reference_the_record"
    if not any(a["signer_has_standing"] for a in referencing):
        return "not_established", "attribution_without_standing"
    return "valid", "attribution_established_by_later_record"


def reference(inp: dict) -> tuple[str, str, str, str]:
    verdict, reason = accountability(inp, check_interval=True, check_multiplicity=True)
    return inp["chain_state"], verdict, reason, anomaly_for(inp)


def n1_authenticated_is_accountable(inp: dict) -> tuple[str, str, str, str]:
    authorized = inp["chain_state"] == "valid"
    return (
        inp["chain_state"],
        "valid" if authorized else "not_established",
        "individual_identity_bound" if authorized else "shared_identity_attribution_pending_or_external",
        anomaly_for(inp),
    )


def n2_shared_flag_only(inp: dict) -> tuple[str, str, str, str]:
    verdict, reason = accountability(inp, check_interval=False, check_multiplicity=False)
    return inp["chain_state"], verdict, reason, anomaly_for(inp)


CONTROLS = {
    "N1-authenticated-is-accountable": n1_authenticated_is_accountable,
    "N2-shared-flag-only": n2_shared_flag_only,
}


def expected_tuple(case: dict) -> tuple[str, str, str, str]:
    e = case["expected"]
    return e["authority_verdict"], e["accountability_verdict"], e["reason"], e["anomaly"]


def main() -> int:
    vectors = json.loads((HERE / "vectors.json").read_text())
    pinned = json.loads((HERE / "input-digests.json").read_text())["inputs"]
    cases = vectors["cases"]

    print(f"{vectors['family']}: {len(cases)} cases, Python runner")

    if vectors["status"] != "candidate_against_proposed":
        fail(f"family status is {vectors['status']!r}")
    if not vectors.get("proposed_text", {}).get("commit"):
        fail("family does not name the proposed text it is a candidate against")

    ids = set()
    for case in cases:
        if case["id"] in ids:
            fail(f"{case['id']}: duplicate vector id")
        ids.add(case["id"])
        if case["label"] != "candidate_against_proposed":
            fail(f"{case['id']}: label is {case['label']!r}")
        if not case["id"].startswith(case["case_id"]):
            fail(f"{case['id']}: vector id does not carry its CASES.md case id")
        if not case.get("proposed_text", {}).get("sections"):
            fail(f"{case['id']}: names no proposed text")
        for field in ("authority_verdict", "accountability_verdict"):
            if case["expected"][field] not in VOCABULARY:
                fail(f"{case['id']}: {field} {case['expected'][field]!r} is outside the settled vocabulary")
        if case["expected"]["anomaly"] not in ANOMALIES:
            fail(f"{case['id']}: unknown anomaly value {case['expected']['anomaly']!r}")
    print(f"  ok   labelling: {len(cases)} cases carry candidate_against_proposed and name their proposed text")

    for case in cases:
        digest = hashlib.sha256(canonicalize_jcs(case["input"]).encode("utf-8")).hexdigest()
        if pinned.get(case["id"]) is None:
            fail(f"{case['id']}: no pinned input digest")
        elif digest != pinned[case["id"]]:
            fail(f"{case['id']}: input digest {digest} does not match pinned {pinned[case['id']]}")
    if len(pinned) != len(cases):
        fail(f"input-digests.json pins {len(pinned)} inputs for {len(cases)} cases")
    print(f"  ok   canonical bytes: {len(cases)}/{len(cases)} JCS input digests match, via PyPI canonicalize_jcs")

    if not all(c["input"]["chain_state"] == "valid" for c in cases):
        fail("a case in this family has a chain that is not valid, which is not what the section is about")
    diverging = [
        c
        for c in cases
        if c["expected"]["authority_verdict"] == "valid"
        and c["expected"]["accountability_verdict"] == "not_established"
    ]
    if not diverging:
        fail("no case pairs a valid chain with accountability not_established, so the two axes never diverge")
    print(
        f"  ok   axis separation: all {len(cases)} chains valid, {len(diverging)} cases pair a valid chain "
        f"with accountability not_established"
    )

    matched = 0
    for case in cases:
        got = reference(case["input"])
        want = expected_tuple(case)
        if got != want:
            fail(f"{case['id']}: reference gave {'/'.join(got)}, expected {'/'.join(want)}")
        else:
            matched += 1
    print(f"  ok   reference model: {matched}/{len(cases)} matched")

    for control in vectors["policies"]["negative_controls"]:
        run = CONTROLS.get(control["id"])
        if run is None:
            fail(f"{control['id']}: no control by that id in this runner")
            continue
        observed = [c["id"] for c in cases if run(c["input"]) != expected_tuple(c)]
        declared = sorted(control["declared_fail_set"])
        if sorted(observed) != declared:
            fail(f"{control['id']}: declared fail set {declared}, observed {sorted(observed)}")
        else:
            print(
                f"  ok   {control['id']}: ran all {len(cases)} cases, "
                f"failed exactly {len(declared)} declared ({', '.join(declared)})"
            )
        if not declared:
            fail(f"{control['id']}: a negative control with an empty declared fail set is not a control")
        if len(declared) == len(cases):
            fail(f"{control['id']}: a control that fails every case has no passing baseline and isolates nothing")

    if problems:
        print(f"\nlifecycle-shared-identity Python: FAILED, {len(problems)} problem(s)", file=sys.stderr)
        return 1
    print(f"\nlifecycle-shared-identity Python: {len(cases)}/{len(cases)} passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: fixtures/lifecycle-shared-identity-with-no-accountable-principal/verify.ts
// Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
//
// TypeScript runner for the lifecycle-shared-identity-with-no-accountable-principal
// candidate family.
//
// Five checks, all of which must pass:
//
//   1. Labelling. Every case carries status candidate_against_proposed, names
//      the proposed text it tests, carries its CASES.md case id in its vector
//      id, and uses only the settled verdict vocabulary for both verdict fields
//      and only a known value for the anomaly flag.
//   2. Canonical bytes. Every case's `input` recanonicalizes under RFC 8785 to
//      its pinned SHA-256, through the published npm package's canonicalizeJCS.
//   3. The authority axis is not the accountability axis. Every case in this
//      family has a valid chain, and at least one case must pair that valid
//      chain with accountability not_established. A family in which the two
//      axes never diverge would not be testing anything the section is about,
//      so this runner refuses to pass one.
//   4. Reference model. The reference evaluator matches each case's expected
//      authority verdict, accountability verdict, reason and anomaly.
//   5. Negative controls. Each control runs against every case, and its
//      observed fail set must equal its declared fail set.
//
// Run: npx tsx fixtures/lifecycle-shared-identity-with-no-accountable-principal/verify.ts
// Exit 0 on success, 1 on any failure.

import { readFileSync } from 'node:fs'
import { createHash } from 'node:crypto'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

import { canonicalizeJCS } from 'agent-passport-system'

import { CONTROLS, evaluateReference, type Input, type Outcome } from './harness.ts'

const HERE = dirname(fileURLToPath(import.meta.url))
const VOCABULARY = new Set(['valid', 'invalid', 'not_established', 'not_yet_effective', 'suspended', 'restricted'])
const ANOMALIES = new Set(['not_applicable', 'on_enumerated_list', 'off_enumerated_list'])

interface Case {
  id: string
  case_id: string
  role: string
  label: string
  proposed_text: { sections: string[] }
  input: Input
  expected: Outcome
}

const vectors = JSON.parse(readFileSync(join(HERE, 'vectors.json'), 'utf8')) as {
  family: string
  status: string
  proposed_text: { repository: string; commit: string }
  policies: { negative_controls: { id: string; declared_fail_set: string[] }[] }
  cases: Case[]
}
const pinned: Record<string, string> = JSON.parse(readFileSync(join(HERE, 'input-digests.json'), 'utf8')).inputs

const failures: string[] = []
const fail = (m: string) => {
  failures.push(m)
  console.error(`  FAIL ${m}`)
}

const same = (a: Outcome, b: Outcome) =>
  a.authority_verdict === b.authority_verdict &&
  a.accountability_verdict === b.accountability_verdict &&
  a.reason === b.reason &&
  a.anomaly === b.anomaly

const render = (o: Outcome) => `${o.authority_verdict}/${o.accountability_verdict}/${o.reason}/${o.anomaly}`

console.log(`${vectors.family}: ${vectors.cases.length} cases, TypeScript runner`)

// --- 1. Labelling -----------------------------------------------------------

if (vectors.status !== 'candidate_against_proposed') fail(`family status is "${vectors.status}"`)
if (!vectors.proposed_text?.repository || !vectors.proposed_text?.commit) {
  fail('family does not name the proposed text it is a candidate against')
}
const seen = new Set<string>()
for (const c of vectors.cases) {
  if (seen.has(c.id)) fail(`${c.id}: duplicate vector id`)
  seen.add(c.id)
  if (c.label !== 'candidate_against_proposed') fail(`${c.id}: label is "${c.label}"`)
  if (!c.case_id || !c.id.startsWith(c.case_id)) fail(`${c.id}: vector id does not carry its CASES.md case id`)
  if (!Array.isArray(c.proposed_text?.sections) || c.proposed_text.sections.length === 0) {
    fail(`${c.id}: names no proposed text`)
  }
  for (const field of ['authority_verdict', 'accountability_verdict'] as const) {
    if (!VOCABULARY.has(c.expected[field])) {
      fail(`${c.id}: ${field} "${c.expected[field]}" is outside the settled vocabulary`)
    }
  }
  if (!ANOMALIES.has(c.expected.anomaly)) fail(`${c.id}: unknown anomaly value "${c.expected.anomaly}"`)
}
console.log(`  ok   labelling: ${vectors.cases.length} cases carry candidate_against_proposed and name their proposed text`)

// --- 2. Canonical bytes -----------------------------------------------------

for (const c of vectors.cases) {
  const got = createHash('sha256').update(Buffer.from(canonicalizeJCS(c.input as never), 'utf8')).digest('hex')
  if (pinned[c.id] === undefined) fail(`${c.id}: no pinned input digest`)
  else if (got !== pinned[c.id]) fail(`${c.id}: input digest ${got} does not match pinned ${pinned[c.id]}`)
}
if (Object.keys(pinned).length !== vectors.cases.length) {
  fail(`input-digests.json pins ${Object.keys(pinned).length} inputs for ${vectors.cases.length} cases`)
}
console.log(`  ok   canonical bytes: ${vectors.cases.length}/${vectors.cases.length} JCS input digests match, via npm canonicalizeJCS`)

// --- 3. The two axes must be capable of diverging ---------------------------

const allChainsValid = vectors.cases.every(c => c.input.chain_state === 'valid')
const diverging = vectors.cases.filter(
  c => c.expected.authority_verdict === 'valid' && c.expected.accountability_verdict === 'not_established',
)
if (!allChainsValid) fail('a case in this family has a chain that is not valid, which is not what the section is about')
if (diverging.length === 0) {
  fail('no case pairs a valid chain with accountability not_established, so the two axes never diverge')
}
console.log(
  `  ok   axis separation: all ${vectors.cases.length} chains valid, ${diverging.length} cases pair a valid chain with accountability not_established`,
)

// --- 4. Reference model -----------------------------------------------------

let matched = 0
for (const c of vectors.cases) {
  const got = evaluateReference(c.input)
  if (!same(got, c.expected)) fail(`${c.id}: reference gave ${render(got)}, expected ${render(c.expected)}`)
  else matched++
}
console.log(`  ok   reference model: ${matched}/${vectors.cases.length} matched`)

// --- 5. Negative controls ---------------------------------------------------

for (const control of vectors.policies.negative_controls) {
  const run = CONTROLS[control.id]
  if (run === undefined) {
    fail(`${control.id}: no control by that id in the harness`)
    continue
  }
  const observed = vectors.cases.filter(c => !same(run(c.input), c.expected)).map(c => c.id)
  const declared = [...control.declared_fail_set].sort()
  if (JSON.stringify(declared) !== JSON.stringify([...observed].sort())) {
    fail(`${control.id}: declared fail set ${JSON.stringify(declared)}, observed ${JSON.stringify(observed.sort())}`)
  } else {
    console.log(
      `  ok   ${control.id}: ran all ${vectors.cases.length} cases, failed exactly ${declared.length} declared (${declared.join(', ')})`,
    )
  }
  if (declared.length === 0) fail(`${control.id}: a negative control with an empty declared fail set is not a control`)
  if (declared.length === vectors.cases.length) {
    fail(`${control.id}: a control that fails every case has no passing baseline and isolates nothing`)
  }
}

if (failures.length > 0) {
  console.error(`\nlifecycle-shared-identity TypeScript: FAILED, ${failures.length} problem(s)`)
  process.exit(1)
}
console.log(`\nlifecycle-shared-identity TypeScript: ${vectors.cases.length}/${vectors.cases.length} passed`)
